### PR TITLE
Use temporary list for storage folder enumeration

### LINKED ISSUE #2878: Duplicate items in SMB drive after "Cache files and folders for better performance" enabled
**Describe the bug**
While "Cache files and folders for better performance" is enabled (seems default on), the file/folder list in a mounted SMB drive will duplicate after every refresh. 

**To Reproduce**
1. Enable "Cache files and folders for better performance"
2. Open a mounted SMB drive
3. refresh

**Screenshots**
屏幕截图 2021-01-07 162642

**Desktop (please complete the following information):**
 - OS Version: Windows 10 21286
 - App version: 1.0.0.0

### PR DESCRIPTION
**Pull request by jakoss**
Fixes #2878

@gave92 I'm not convinced about the enumeration method right now. Enumerating via file system one by one seems completly unnecessary, and i believe it's inefficient. Is this by design?
image

I believe that it would be the best to create load request for all items or pretty large batches. Operating system should take care about the caching etc.

That being said - are you open for Pull Requests with refactoring? I'm pretty sure i'll have to do some refactoring to implement preemptive caching anyway

### CODE REVIEW

**Comment by gave92** [comment]
> I'm not convinced about the enumeration method right now

I'm not convinced it's the best way either. This was done to:
- Immediately show results to user -> can be done anyway by making the first batch "small" and the rest larger
- Handle the UnauthorizedAccessException and FileNotFoundException exceptions -> this check was added after seeing bug reports but I was unable to reproduce. What happens if you ask GetItemsAsync for 100 items and only one throws UnauthorizedAccessException?

**Comment by jakoss** [comment]
The solution might be to do 1by1 enumeration on Exception encounter on batch call, making it optional only for those cases. I'll deal with that during preemptive cache refactoring if you don't mind

File: Files/ViewModels/ItemViewModel.cs [hunk 1]
```diff
@@ -1227,13 +1228,15 @@ private async Task EnumFromStorageFolderAsync()
                 }
                 if (item.IsOfType(StorageItemTypes.Folder))
                 {
-                    await AddFolderAsync(item as StorageFolder, returnformat);
+                    var folder = await AddFolderAsync(item as StorageFolder, returnformat);
+                    tempList.Add(folder);
```
Comment (gave92) [inline Files/ViewModels/ItemViewModel.cs:1232]: Add: `if (folder != null) tempList.Add(folder);`

File: Files/ViewModels/ItemViewModel.cs [hunk 2]
```diff
@@ -1227,13 +1228,15 @@ private async Task EnumFromStorageFolderAsync()
...
                     ++count;
                 }
                 else
                 {
                     var file = item as StorageFile;
-                    await AddFileAsync(file, returnformat, true);
+                    var fileEntry = await AddFileAsync(file, returnformat, true);
+                    tempList.Add(fileEntry);
```
Comment (gave92) [inline Files/ViewModels/ItemViewModel.cs:1239]: Add: if (fileEntry != null) tempList.Add(fileEntry);

File: Files/ViewModels/ItemViewModel.cs [hunk 3]
```diff
@@ -1242,14 +1245,23 @@ private async Task EnumFromStorageFolderAsync()
                 }
                 if (count % 300 == 0)
```
Comment (gave92) [inline Files/ViewModels/ItemViewModel.cs:1246]: Maybe do the same as standard enumeration: `count == 32 || count % 300 == 0` so the user starts seeing items sooner

File: Files/ViewModels/ItemViewModel.cs [hunk 4]
```diff
@@ -1242,14 +1245,23 @@ private async Task EnumFromStorageFolderAsync()
...
                 {
-                    OrderFiles();
+                    var orderedList = OrderFiles2(tempList);
+                    await CoreApplication.MainView.ExecuteOnUIThreadAsync(() =>
```
Comment (gave92) [inline Files/ViewModels/ItemViewModel.cs:1249]: This should be unnecessary, enumeration from storagefolder is done on UI thread

File: Files/ViewModels/ItemViewModel.cs [hunk 5]
```diff
@@ -1878,17 +1890,13 @@ public void AddItemsToCollectionAsync(string path, string previousDir)
             RapidAddItemsToCollectionAsync(path, previousDir);
         }
 
-        public async Task AddFolderAsync(StorageFolder folder, string dateReturnFormat)
+        public async Task<ListedItem> AddFolderAsync(StorageFolder folder, string dateReturnFormat)
         {
             var basicProperties = await folder.GetBasicPropertiesAsync();
 
-            if ((AssociatedInstance.ContentFrame.SourcePageType == typeof(GenericFileBrowser)) || (AssociatedInstance.ContentFrame.SourcePageType == typeof(GridViewBrowser)))
+            if (!_addFilesCTS.IsCancellationRequested && ((AssociatedInstance.ContentFrame.SourcePageType == typeof(GenericFileBrowser)) || (AssociatedInstance.ContentFrame.SourcePageType == typeof(GridViewBrowser))))
```
Comment (gave92) [inline Files/ViewModels/ItemViewModel.cs:1897]: Could you write this check as (personal preference && coherence with other similar checks):
```cs
if (_addFilesCTS.IsCancellationRequested)
{	
     return null;	
}
if ((AssociatedInstance.ContentFrame.SourcePageType == typeof(GenericFileBrowser)) || (AssociatedInstance.ContentFrame.SourcePageType == typeof(GridViewBrowser)))
{
    return new ....
}
```

**Review by gave92 — CHANGES_REQUESTED** [review]

File: Files/ViewModels/ItemViewModel.cs [hunk 6]
```diff
@@ -1904,11 +1912,12 @@ public async Task AddFolderAsync(StorageFolder folder, string dateReturnFormat)
...
                     FileSizeBytes = 0
                     //FolderTooltipText = tooltipString,
-                });
+                };
             }
+            return null;
         }
 
-        public async Task AddFileAsync(StorageFile file, string dateReturnFormat, bool suppressThumbnailLoading = false)
+        public async Task<ListedItem> AddFileAsync(StorageFile file, string dateReturnFormat, bool suppressThumbnailLoading = false)
```
Comment (gave92) [inline Files/ViewModels/ItemViewModel.cs:1920]: Could you make this `private`?

File: Files/ViewModels/ItemViewModel.cs [hunk 7]
```diff
@@ -1878,17 +1890,13 @@ public void AddItemsToCollectionAsync(string path, string previousDir)
             RapidAddItemsToCollectionAsync(path, previousDir);
         }
 
-        public async Task AddFolderAsync(StorageFolder folder, string dateReturnFormat)
+        public async Task<ListedItem> AddFolderAsync(StorageFolder folder, string dateReturnFormat)
```
Comment (gave92) [inline Files/ViewModels/ItemViewModel.cs:1893]: Could you make this `private`?

**Review by gave92 — APPROVED** [review]

**Review by gave92 — APPROVED** [review]
Looks good. Thanks a lot 👍

**Comment by yaichenbaum** [comment]
@jakoss How will this effect the time it takes to load large directoreis when the setting to cache items is turned off?

**Comment by jakoss** [comment]
It still flushes files list to UI when loaded 32 files or when `count % 300 == 0`. So basically it works the same as the native loading right now

**Comment by jakoss** [comment]
Also, until now, i believe that items loaded with storage enumeration weren't properly ordered until count of 300. There was no Order after finished loading, so smaller lists might be not ordered. But i'm not 100% sure about that

**Comment by gave92** [comment]
@jakoss i think you're right. Except GetItemsAsync() may already be returning items sorted by name by default (need to verify this). In any case the changes make it work even if the user changes the sort mode/direction

**Comment by yaichenbaum** [comment]
@jakoss Looks good then. By the way, the issue where directories "load" twice when using the cache options seems to be happening consistently now.

**Comment by jakoss** [comment]
That might be the issues with updating the observable collection of files, but i didn't observe that today once. I'll try and dig more about that later on

**Pull request merged**